### ISSUE TITLE
Build failure on FreeBSD

### ISSUE THREAD

**Issue by nkosi23** [mobilecoinofficial/full-service#980]
**Describe the bug**
When building the full service API on FreeBSD, everything builds smoothly until the 798th package out of 924, the build error is:

**To Reproduce**

Install rustup on FreeBSD as explained in: https://gist.github.com/fabiotatsuo/e88b025baf3dfa47683419948723706f

Install the build dependencies with the command:

pkg install cmake protobuf llvm llvm-devel sqlite3 sqlcipher git bash

**Expected behavior**
The build completes without error. Rusts binaries are usually easily portable to FreeBSD, it looks like this issue could just be a matter of adding a few config files to support FreeBSD.

**Environment (please complete the following information):**
 - OS: FreeBSD 
 - Version 13

**<details><summary>Logs</summary>**

```
error: failed to run custom build command for `hidapi v2.4.1`

Caused by:
  process didn't exit successfully: `/home/mobilecoin/full-service/target/release/build/hidapi-0f1b6d5217db2a16/build-script-build` (exit status: 101)
  --- stdout
  cargo:rerun-if-env-changed=HIDAPI_NO_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG_x86_64-unknown-freebsd
  cargo:rerun-if-env-changed=PKG_CONFIG_x86_64_unknown_freebsd
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG
  cargo:rerun-if-env-changed=HIDAPI_STATIC
  cargo:rerun-if-env-changed=HIDAPI_DYNAMIC
  cargo:rerun-if-env-changed=PKG_CONFIG_ALL_STATIC
  cargo:rerun-if-env-changed=PKG_CONFIG_ALL_DYNAMIC
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64-unknown-freebsd
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64_unknown_freebsd
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64-unknown-freebsd
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64_unknown_freebsd
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64-unknown-freebsd
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64_unknown_freebsd
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=HIDAPI_STATIC
  cargo:rerun-if-env-changed=HIDAPI_DYNAMIC
  cargo:rerun-if-env-changed=PKG_CONFIG_ALL_STATIC
  cargo:rerun-if-env-changed=PKG_CONFIG_ALL_DYNAMIC
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64-unknown-freebsd
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64_unknown_freebsd
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64-unknown-freebsd
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64_unknown_freebsd
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64-unknown-freebsd
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64_unknown_freebsd
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR

  --- stderr
  thread 'main' panicked at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/hidapi-2.4.1/build.rs:131:41:
  Unable to find hidapi: Could not run `PKG_CONFIG_ALLOW_SYSTEM_LIBS=1 PKG_CONFIG_ALLOW_SYSTEM_CFLAGS=1 pkg-config --libs --cflags hidapi`
  The pkg-config command could not be found.

  Most likely, you need to install a pkg-config package for your OS.
  Try `apt install pkg-config`, or `yum install pkg-config`,
  or `pkg install pkg-config`, or `apk add pkgconfig` depending on your distribution.

  If you've already installed it, ensure the pkg-config command is one of the
  directories in the PATH environment variable.

  If you did not expect this build to link to a pre-installed system library,
  then check documentation of the hidapi crate for an option to
  build the library from source, or disable features or dependencies
  that require pkg-config.
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

```

And the build eventually ends with [these logs](https://github.com/mobilecoinofficial/full-service/files/15340384/mobilecoin-build-log.txt).

</details>

**Additional context**

**Comment by nick-mobilecoin** [comment]
if you want FreeBSD native you might be able to get by using https://github.com/mobilecoinofficial/full-service/blob/main/tools/build-fs.sh
```
./tools/build-fs.sh main
```

The error your getting is because the enclave css(not cascading style sheets) environment variables are not specified via:
- `CONSENSUS_ENCLAVE_CSS`
- `INGETS_ENCLAVE_CSS`

When these aren't specified then the build tries to build the enclaves to generate the css files. 
However the enclave builds are limited to x86_64 linux.

The `build-fs.sh` will download the current css files for you and then build full service using them.

you could also use the docker image. The `mob` command will invoke docker for the appropriate image.
```
./mob prompt
cargo build
```

**Comment by nick-mobilecoin** [comment]
> Any chance to put it behind a platform-specific compile flag 

I did a little digging around, it looks like the ledger logic is spread around a bit in the code base.
So it would likely take a bit to isolate the ledger code.

My vote would be that it goes behind a feature flag maybe something like `hw_signing` or similar. 
Unfortunately I don't think this will get prioritized from our end at this time.

**Comment by holtzman** [comment]
as a quick experiment, would you see if the experiment/no-hardware-wallet branch builds on freebsd?  It's building on MacOS sans the ledger-mob submodule, though I haven't really tested it.

As Nick said, it would be hard for us to prioritize refactoring things right now, so just gathering data...

**Comment by holtzman** [comment]
@nkosi23 I have removed the bug label as full-service is not supported for freebsd.

**Comment by holtzman** [comment]
@nkosi23, fyi, I created a freebsd 14 instance on EC2 and it built grpcio-sys. Perhaps there are packages in freebsd 13 that could be upgraded to get you past that?

Now I'm stuck at linking.  There appear to be encryption-related symbols that sq3lite is using but the linker is not finding. Full-service will optionally encrypt the wallet-db...

```
          >>> referenced by sqlite3.c:106013 (sqlcipher/sqlite3.c:106013)
          >>>               sqlite3.o:(sqlcipher_openssl_hmac) in archive /home/ec2-user/dev/full-service/target/release/deps/liblibsqlite3_sys-b42aee77f4cad872.rlib

          ld: error: undefined symbol: EVP_MAC_CTX_new
          >>> referenced by sqlite3.c:106020 (sqlcipher/sqlite3.c:106020)
          >>>               sqlite3.o:(sqlcipher_openssl_hmac) in archive /home/ec2-user/dev/full-service/target/release/deps/liblibsqlite3_sys-b42aee77f4cad872.rlib

          ld: error: undefined symbol: EVP_MAC_init
          >>> referenced by sqlite3.c:106029 (sqlcipher/sqlite3.c:106029)
          >>>               sqlite3.o:(sqlcipher_openssl_hmac) in archive /home/ec2-user/dev/full-service/target/release/deps/liblibsqlite3_sys-b42aee77f4cad872.rlib
          >>> referenced by sqlite3.c:106043 (sqlcipher/sqlite3.c:106043)
          >>>               sqlite3.o:(sqlcipher_openssl_hmac) in archive /home/ec2-user/dev/full-service/target/release/deps/liblibsqlite3_sys-b42aee77f4cad872.rlib
          >>> referenced by sqlite3.c:106036 (sqlcipher/sqlite3.c:106036)
          >>>               sqlite3.o:(sqlcipher_openssl_hmac) in archive /home/ec2-user/dev/full-service/target/release/deps/liblibsqlite3_sys-b42aee77f4cad872.rlib

          ld: error: undefined symbol: EVP_MAC_update
          >>> referenced by sqlite3.c:106054 (sqlcipher/sqlite3.c:106054)
          >>>               sqlite3.o:(sqlcipher_openssl_hmac) in archive /home/ec2-user/dev/full-service/target/release/deps/liblibsqlite3_sys-b42aee77f4cad872.rlib
          >>> referenced by sqlite3.c:106061 (sqlcipher/sqlite3.c:106061)
          >>>               sqlite3.o:(sqlcipher_openssl_hmac) in archive /home/ec2-user/dev/full-service/target/release/deps/liblibsqlite3_sys-b42aee77f4cad872.rlib

          ld: error: undefined symbol: EVP_MAC_final
          >>> referenced by sqlite3.c:106068 (sqlcipher/sqlite3.c:106068)
          >>>               sqlite3.o:(sqlcipher_openssl_hmac) in archive /home/ec2-user/dev/full-service/target/release/deps/liblibsqlite3_sys-b42aee77f4cad872.rlib
          >>> referenced by sqlite3.c:106074 (sqlcipher/sqlite3.c:106074)
          >>>               sqlite3.o:(sqlcipher_openssl_hmac) in archive /home/ec2-user/dev/full-service/target/release/deps/liblibsqlite3_sys-b42aee77f4cad872.rlib

          ld: error: undefined symbol: EVP_MAC_CTX_free
          >>> referenced by sqlite3.c:106087 (sqlcipher/sqlite3.c:106087)
          >>>               sqlite3.o:(sqlcipher_openssl_hmac) in archive /home/ec2-user/dev/full-service/target/release/deps/liblibsqlite3_sys-b42aee77f4cad872.rlib

          ld: error: undefined symbol: EVP_MAC_free
          >>> referenced by sqlite3.c:106088 (sqlcipher/sqlite3.c:106088)
          >>>               sqlite3.o:(sqlcipher_openssl_hmac) in archive /home/ec2-user/dev/full-service/target/release/deps/liblibsqlite3_sys-b42aee77f4cad872.rlib

          ld: error: undefined symbol: EVP_CIPHER_get_nid
          >>> referenced by sqlite3.c:106186 (sqlcipher/sqlite3.c:106186)
          >>>               sqlite3.o:(sqlcipher_openssl_get_cipher) in archive /home/ec2-user/dev/full-service/target/release/deps/liblibsqlite3_sys-b42aee77f4cad872.rlib

          ld: error: undefined symbol: EVP_CIPHER_get_key_length
          >>> referenced by sqlite3.c:106190 (sqlcipher/sqlite3.c:106190)
          >>>               sqlite3.o:(sqlcipher_openssl_get_key_sz) in archive /home/ec2-user/dev/full-service/target/release/deps/liblibsqlite3_sys-b42aee77f4cad872.rlib

          ld: error: undefined symbol: EVP_CIPHER_get_iv_length
          >>> referenced by sqlite3.c:106194 (sqlcipher/sqlite3.c:106194)
          >>>               sqlite3.o:(sqlcipher_openssl_get_iv_sz) in archive /home/ec2-user/dev/full-service/target/release/deps/liblibsqlite3_sys-b42aee77f4cad872.rlib

          ld: error: undefined symbol: EVP_CIPHER_get_block_size
          >>> referenced by sqlite3.c:106198 (sqlcipher/sqlite3.c:106198)
          >>>               sqlite3.o:(sqlcipher_openssl_get_block_sz) in archive /home/ec2-user/dev/full-service/target/release/deps/liblibsqlite3_sys-b42aee77f4cad872.rlib

          ld: error: undefined symbol: EVP_MD_get_size
          >>> referenced by sqlite3.c:0 (sqlcipher/sqlite3.c:0)
          >>>               sqlite3.o:(sqlcipher_openssl_get_hmac_sz) in archive /home/ec2-user/dev/full-service/target/release/deps/liblibsqlite3_sys-b42aee77f4cad872.rlib
          >>> referenced by sqlite3.c:0 (sqlcipher/sqlite3.c:0)
          >>>               sqlite3.o:(sqlcipher_openssl_get_hmac_sz) in archive /home/ec2-user/dev/full-service/target/release/deps/liblibsqlite3_sys-b42aee77f4cad872.rlib
          >>> referenced by sqlite3.c:0 (sqlcipher/sqlite3.c:0)
          >>>               sqlite3.o:(sqlcipher_openssl_get_hmac_sz) in archive /home/ec2-user/dev/full-service/target/release/deps/liblibsqlite3_sys-b42aee77f4cad872.rlib
          cc: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by holtzman** [comment]
Good question!

TL; DR: full-service has few mandatory updates and they are far between. One is imminent, but we forked the experimental branch off main *after* the mandatory updates had landed, so that branch is already good to go.

When full-service attests the enclaves it uses, it does not require that their individual build measurements are on a whitelist compiled into full-service. This is exactly so that the same version of full-service can continue to work across enclave-side updates, as long as there are no breaking protocol changes.  It does validate that the enclave was signed by the trusted authority, has the expected product id, and meets a minimum version requirement before trusting it.

In contrast, Sentz and Signal do validate that each enclave's build measurements are on a compiled-in whitelist.  The environments of use, ease of updating, and education / contextual awareness level of the operator of each of these are very different and thus different policies are appropriate.

On occasion (3 times so far) we have had to make breaking changes in the protocol that have required full-service to be updated. The first was when we went multi-asset (~10/22); the second was phase 2 of multi-asset which introduced signed-contingent-inputs, allowing for atomic swaps between assets (~3/2023); and, the third will be to switch from SGX EPID attestation to SGX DCAP due to Intel phasing out EPID support (7/2024).

The timing of this branch is very good for that 3rd one as we are mid-way through the deployment process, and full-service is already updated and compatible with both the current and soon-to-be-deployed enclaves. In fact, testnet is already upgraded to v6.0, which is scheduled to be deployed to mainnet the week of July 8.

**Comment by holtzman** [comment]
Inspired by the topic, I merged main back into this branch as we've done a bunch of feature updates since the branch forked.

I built and tested on MacOS, including a scripted client test that end-to-end exercises a running full-service via its api.

**Comment by nick-mobilecoin** [comment]
Thanks for the heads up, looks like testnet node1 has been having some issues since about Oct 8th.
one can see it's signature is missing from the block explorer starting at, [3193708](https://block-explorer.test.mobilecoin.com/blocks/3193708)
We're currently looking into it
